### ISSUE TITLE
flexattr - import from and write to file metadata

### ISSUE THREAD

**Issue by enyg** [beetbox/beets#3326]
### Use case

I'd like to have user defined tags that already exist in a track's metadata imported to the beets library. Currently, it appears that beets will import any metadata fields that correspond to the built-in fields, but ignores other metadata fields present in the files.

Additionally, `write` does not write any flexattrs that have been created in the beets library back to the files.

Flexattrs could potentially be much more powerful if the changes made with beets could be read by external music players. Using the metadata in the files themselves seems like the most portable way to do this. Any fields specific to a music player could be manipulated with beets as well.

### Solution
My understanding is that Vorbis comments allow for arbitrary field names, so I think it would be straightforward to map flexattr keys to vorbis keys.

ID3v2.4 has TXXX frames (User defined text information frame), which can provide similar functionality. See [http://id3.org/id3v2.4.0-frames](http://id3.org/id3v2.4.0-frames)

If beets can already do this, I could not see a way to do it.

**Comment by sampsyo** [comment]
Interesting idea! For what it's worth, it's currently possible (although quite hacky) for plugins to map _individual_ flexible attributes to specific tag storage formats:
http://docs.beets.io/en/stable/dev/plugins.html#extend-mediafile

But the way I read this feature request is for some _bulk_ option to map _all_ flexible attributes to on-disk metadata using the same name. Clearly, this will need some fairly fancy support from MediaFile before beets can support it. I think the biggest challenge will be managing conflicts with built-in metadata, which can't necessarily be abstracted using a 1:1 mapping to files' tags. One option would be to "namespace" all the flexattrs' tag mappings so they can't possibly conflict (e.g., by prefixing the name with "beets:" or something), but I gather that probably wouldn't meet the goal of interoperability with other tools.

**Comment by enyg** [comment]
You're right that the namespacing method wouldn't solve the problem, at least for the use cases I have in mind. 

Would it be feasible to have a finite list of special cases which are mapped to the built-in metadata (meaning the non-flexible attributes such as Artist, Album, Genre, etc.) and map any key that doesn't match a built-in field to a custom key of the same name? Maybe I'm missing where some of the mapping conflicts would come into play.

If you've ever used foobar2000, this functionality is there. (In fact, that's probably the only program I've seen that does it.) It's not open source though, unfortunately.

**Comment by sampsyo** [comment]
Something like that might be possible, but it's made more complicated by the fact that the list of built-in tags (with proper mappings) changes over time—namely, we add new tags pretty frequently. So excluding them with a list would cause some flexible attributes to surprisingly "disappear" when we add support for a new tag. Perhaps that's an acceptable inconvenience, but it would be pretty weird.

I think it's also worth considering that, even in this hypothetical mode, you might want to have some fields that are "database-only" fields that don't have a representation in the files. These could be fields that change too frequently to be practical, such as the "play count" field generated by the `mpdstats` plugin. With that in mind, maybe one would want to namespace the beets fields that gain the automatically-map-to-tags superpower.

So I can understand more completely, you mentioned that the goal here would be interop—is foobar2000 the main tool you're thinking of interoperating with? Or are there other hypothetical interop situations you have in mind?

**Comment by kifty** [comment]
Would it be possible to explicitly identify certain mappings in the config file? In my case, I've carefully managed the <RATING> field in foobar2000, and I'd love to bring it over to Beets.

I'm not (much of) a coder, but if a plugin to map specific vorbis keys or TXXX frames to flexible attributes is feasible, I'd give it a try. I'm even happy with a one-time copy on import, if that's significantly easier.
